lib/uuid/ecto/type_config.ex

```diff
@@ -0,0 +1,134 @@
+          |> Keyword.merge(opts)
+          |> keyword_pop!(:type)
+        else
+          keyword_pop!(opts, :type)
+        end
+
+      args = get_and_validate_args(uuid_type, opts)
+
+      {uuid_type, args}
+    end
+
+    @doc """
+    Validate the provided opts for different types.
+    """
+    @spec get_and_validate_args(uuid_type, keyword) :: uuid_type_args
+    def get_and_validate_args(:uuid1, []), do: []
+
+    def get_and_validate_args(:uuid3, opts) do
+      namespace = Keyword.fetch!(opts, :namespace)
+      name = Keyword.fetch!(opts, :name)
+
+      if namespace not in @valid_v3_namespaces do
+        # If the namespace is not in the list of valid namespaces, the only
+        # other valid option is that it is a UUID.
+        unless UUID.valid?(namespace) do
+          namespace_str = Enum.join(@valid_v3_namespaces, "|")
+          raise ArgumentError, message: "Invalid namespace; expected #{namespace_str} or a UUID"
+        end
+      end
+
+      if not is_binary(name) do
+        raise ArgumentError, message: "Invalid name: #{inspect(name)}; expected String"
+      end
+
+      [namespace, name]
+    end
+
+    def get_and_validate_args(:uuid4, []), do: []
+
+    def get_and_validate_args(:uuid5, opts) do
+      namespace = Keyword.fetch!(opts, :namespace)
+      name = Keyword.fetch!(opts, :name)
+
+      if namespace not in @valid_v5_namespaces do
+        # If the namespace is not in the list of valid namespaces, the only
+        # other valid option is that it is a UUID.
+        unless UUID.valid?(namespace) do
+          namespace_str = Enum.join(@valid_v5_namespaces, "|")
+          raise ArgumentError, message: "Invalid namespace; expected #{namespace_str} or a UUID"
+        end
+      end
+
+      if not is_binary(name) do
+        raise ArgumentError, message: "Invalid name: #{inspect(name)}; expected String"
+      end
+
+      [namespace, name]
+    end
+
+    def get_and_validate_args(:uuid6, opts) do
+      node_type = Keyword.get(opts, :node_type, @default_v6_node_type)
+
+      if node_type not in @valid_v6_node_types do
+        node_types_str = Enum.join(@valid_v6_node_types, "|")
+
+        raise ArgumentError,
+          message: "Invalid node type: #{inspect(node_type)}; expected one of #{node_types_str}"
+      end
+
+      [node_type]
+    end
+
+    # Catch-all validator.
+    def get_and_validate_args(type, opts) do
+      raise ArgumentError,
+        message:
+          "Invalid type, or unrecognized option; type: #{inspect(type)}; options: #{inspect(opts)}"
+    end
+
+    # Backfill for Elixir < 1.10.x
+    defp keyword_pop!(keywords, key) when is_list(keywords) and is_atom(key) do
+      if function_exported?(Keyword, :pop!, 2) do
+        Keyword.pop!(keywords, key)
+      else
+        case Keyword.fetch(keywords, key) do
+          {:ok, value} -> {value, Keyword.delete(keywords, key)}
+          :error -> raise KeyError, key: key, term: keywords
+        end
+      end
+    end
```

Comment:
nitpick: it's more efficient to perform the check at compile time (but the difference is negligible)

```suggestion
   if function_exported?(Keyword, :pop!, 2) do
      defp keyword_pop!(keywords, key) when is_list(keywords) and is_atom(key) do
        Keyword.pop!(keywords, key)
      end
    else
      defp keyword_pop!(keywords, key) when is_list(keywords) and is_atom(key) do
        case Keyword.fetch(keywords, key) do
          {:ok, value} -> {value, Keyword.delete(keywords, key)}
          :error -> raise KeyError, key: key, term: keywords
        end
      end
    end
```